### ISSUE TITLE
FirestoreDataSource - Please add the ability to get an occurred exception

### ISSUE THREAD

**Issue by Zeliret** [firebase/FirebaseUI-Android#1592]
Right now I can only get an error status and call retry. But I need to get the exact exception to provide an explanation to my users.
I have found the place where the error state is populated and It seems that I need to override a lot of code to get the error message. 
Is there a good way to workaround this?

**Comment by samtstern** [comment]
@Zeliret can you show an example of the code you're using now and maybe some comments on how you'd want it to improve?

**Comment by Zeliret** [comment]
@samtstern sure. Thanks for the answer.
So, I have a Query instance and configured PagedList builder.
I create a new adapter:
```kotlin
val options = FirestorePagingOptions.Builder<Trip>()
                        .setLifecycleOwner(this)
                        .setQuery(result.data.query, Source.DEFAULT, pagingConfig) {
                            result.data.converter(it)
                        }
                        .build()

                    binding.tripsList.swapAdapter(TripsAdapter(options), false)
```
Let's assume I get an error while I'm loading a page. I can check it here:
```kotlin
override fun onLoadingStateChanged(state: LoadingState) {
            if( state == LoadingState.ERROR ) {
                retry()
            }
        }
```
But I can get only the flag. How can I get the actual error? As far as I can see from the source code:
```java
/**
     * Error listener that logs, sets the error state, and sets up retry.
     */
    private abstract class OnLoadFailureListener implements OnFailureListener {

        @Override
        public void onFailure(@NonNull Exception e) {
            Log.w(TAG, "load:onFailure", e);

            // On error we do NOT post any value to the PagedList, we just tell
            // the developer that we are now in the error state.
            mLoadingState.postValue(LoadingState.ERROR);

            // Set the retry action
            mRetryRunnable = getRetryRunnable();
        }

        protected abstract Runnable getRetryRunnable();
    }
```
Here the exception just dumped into the log. 
So I can not check if it is a network error or show it to the user.

**Comment by samtstern** [comment]
@Zeliret ah I see that's a good feature request!  You're right, there's no good way to get it right now in the Paginated adapters.

**Comment by samtstern** [comment]
How would you feel about something like a `getLastError()` call in the paginated adapter classes?

**Comment by Zeliret** [comment]
> How would you feel about something like a `getLastError()` call in the paginated adapter classes?

That sounds good. 
So I could use the flow:
```kotlin
if( state == LoadingState.ERROR ) {
     val ex = getLastError()
     if( ex is IOException )
         retry()
     else
         showErrorMessage(ex)
}
```
As for me that should works. Thank you! :)

**Comment by samtstern** [comment]
Yep that's how I imagine it too.

**Comment by abhinavpramod85** [comment]
can u tell me code to create a law based app

**Comment by PatilShreyas** [comment]
We should solve this issue together as same as we implemented in `FirebaseRecyclerPagingAdapter`. As #1603

**Comment by samtstern** [comment]
This has been fixed and released in `5.0.0`
